File: web/api/tests/test_quality.py
"""
Water Treatment Controller - Data Quality Tests
Copyright (C) 2024
SPDX-License-Identifier: GPL-3.0-or-later

Tests for quality state handling and propagation.
"""

import pytest
from datetime import datetime, timezone, timedelta


class TestQualityCodes:
    """Tests for quality code values and semantics."""

    def test_quality_codes_match_opc_ua(self):
        """Quality codes should match OPC UA specification."""
        # These values are defined in shm_client.py
        QUALITY_GOOD = 0x00
        QUALITY_UNCERTAIN = 0x40
        QUALITY_BAD = 0x80
        QUALITY_NOT_CONNECTED = 0xC0

        assert QUALITY_GOOD == 0x00
        assert QUALITY_UNCERTAIN == 0x40
        assert QUALITY_BAD == 0x80
        assert QUALITY_NOT_CONNECTED == 0xC0

    def test_quality_names_mapping(self):
        """Quality names should be human-readable."""
        QUALITY_NAMES = {
            0x00: "good",
            0x40: "uncertain",
            0x80: "bad",
            0xC0: "not_connected"
        }

        assert QUALITY_NAMES[0x00] == "good"
        assert QUALITY_NAMES[0x40] == "uncertain"
        assert QUALITY_NAMES[0x80] == "bad"
        assert QUALITY_NAMES[0xC0] == "not_connected"


class TestQualityUsability:
    """Tests for quality-based control decisions."""

    def test_good_quality_usable_for_control(self):
        """GOOD quality data should be usable for control."""
        quality = 0x00  # GOOD
        usable = quality in (0x00, 0x40)  # GOOD or UNCERTAIN
        assert usable is True

    def test_uncertain_quality_usable_with_warning(self):
        """UNCERTAIN quality data should be usable but with warning."""
        quality = 0x40  # UNCERTAIN
        usable = quality in (0x00, 0x40)
        should_warn = quality == 0x40
        assert usable is True
        assert should_warn is True

    def test_bad_quality_not_usable(self):
        """BAD quality data should not be used for control."""
        quality = 0x80  # BAD
        usable = quality in (0x00, 0x40)
        assert usable is False

    def test_not_connected_not_usable(self):
        """NOT_CONNECTED quality data should not be used for control."""
        quality = 0xC0  # NOT_CONNECTED
        usable = quality in (0x00, 0x40)
        assert usable is False


class TestStalenessDetection:
    """Tests for staleness detection and quality degradation."""

    def test_fresh_data_is_good(self):
        """Data less than 5 seconds old should remain GOOD."""
        now = datetime.now(timezone.utc)
        timestamp = now - timedelta(seconds=3)
        age = (now - timestamp).total_seconds()

        quality = 0x00  # GOOD
        stale = age > 5

        assert stale is False
        assert quality == 0x00

    def test_stale_data_becomes_uncertain(self):
        """Data 5-60 seconds old should become UNCERTAIN."""
        now = datetime.now(timezone.utc)
        timestamp = now - timedelta(seconds=30)
        age = (now - timestamp).total_seconds()

        quality = 0x00  # Initially GOOD
        stale = False

        if age > 5:
            stale = True
            if quality == 0x00:
                quality = 0x40  # UNCERTAIN

        assert stale is True
        assert quality == 0x40

    def test_extended_stale_becomes_bad(self):
        """Data 60+ seconds old should become BAD."""
        now = datetime.now(timezone.utc)
        timestamp = now - timedelta(seconds=90)
        age = (now - timestamp).total_seconds()

        quality = 0x00  # Initially GOOD

        if age > 60:
            quality = 0x80  # BAD

        assert quality == 0x80

    def test_very_stale_becomes_not_connected(self):
        """Data 5+ minutes old should become NOT_CONNECTED."""
        now = datetime.now(timezone.utc)
        timestamp = now - timedelta(minutes=6)
        age = (now - timestamp).total_seconds()

        quality = 0x00  # Initially GOOD

        if age > 300:  # 5 minutes
            quality = 0xC0  # NOT_CONNECTED

        assert quality == 0xC0


class TestInterlockFailSafe:
    """Tests for interlock fail-safe behavior on bad quality."""

    def test_interlock_trips_on_bad_quality(self):
        """Interlock should trip when input quality is BAD."""
        value = 50.0  # Normal value
        quality = 0x80  # BAD
        threshold = 100.0

        input_valid = quality == 0x00  # Only GOOD is valid for interlocks

        if input_valid:
            condition_met = value > threshold
        else:
            condition_met = True  # Fail-safe

        assert input_valid is False
        assert condition_met is True  # Should trip despite normal value

    def test_interlock_normal_on_good_quality(self):
        """Interlock should evaluate normally with GOOD quality."""
        value = 50.0  # Normal value
        quality = 0x00  # GOOD
        threshold = 100.0

        input_valid = quality == 0x00

        if input_valid:
            condition_met = value > threshold
        else:
            condition_met = True

        assert input_valid is True
        assert condition_met is False  # Should NOT trip


class TestPIDQualityCheck:
    """Tests for PID loop quality checking."""

    def test_pid_accepts_good_quality(self):
        """PID should accept GOOD quality input."""
        iops_status = 0x80  # IOPS_GOOD
        quality = 0x00  # QUALITY_GOOD

        quality_ok = (iops_status == 0x80 and
                      quality in (0x00, 0x40))

        assert quality_ok is True

    def test_pid_accepts_uncertain_quality(self):
        """PID should accept UNCERTAIN quality input with logging."""
        iops_status = 0x80  # IOPS_GOOD
        quality = 0x40  # QUALITY_UNCERTAIN

        quality_ok = (iops_status == 0x80 and
                      quality in (0x00, 0x40))

        should_log_warning = quality == 0x40

        assert quality_ok is True
        assert should_log_warning is True

    def test_pid_rejects_bad_quality(self):
        """PID should reject BAD quality input."""
        iops_status = 0x80  # IOPS_GOOD
        quality = 0x80  # QUALITY_BAD

        quality_ok = (iops_status == 0x80 and
                      quality in (0x00, 0x40))

        assert quality_ok is False

    def test_pid_rejects_bad_iops(self):
        """PID should reject input with bad IOPS status."""
        iops_status = 0x00  # IOPS_BAD
        quality = 0x00  # QUALITY_GOOD

        quality_ok = (iops_status == 0x80 and
                      quality in (0x00, 0x40))

        assert quality_ok is False


class TestHistorianQuality:
    """Tests for historian quality storage."""

    def test_historian_sample_includes_quality(self):
        """Historian samples should include quality field."""
        sample = {
            "timestamp_ms": 1000000,
            "tag_id": 1,
            "value": 7.5,
            "quality": 0x00
        }

        assert "quality" in sample
        assert sample["quality"] == 0x00

    def test_historian_preserves_quality_from_sensor(self):
        """Historian should preserve quality from sensor data."""
        sensor_quality = 0x40  # UNCERTAIN

        sample = {
            "timestamp_ms": 1000000,
            "tag_id": 1,
            "value": 7.5,
            "quality": sensor_quality  # Direct propagation
        }

        assert sample["quality"] == 0x40

File: tests/test_quality.c
/**
 * Water Treatment Controller - Data Quality Tests
 * Copyright (C) 2024
 * SPDX-License-Identifier: GPL-3.0-or-later
 *
 * Tests for quality propagation and fail-safe behavior.
 * These tests verify that BAD quality data causes appropriate
 * system responses (interlock trips, PID holds, etc.)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <assert.h>
#include "../src/types.h"

/* Test counters */
static int tests_run = 0;
static int tests_passed = 0;

#define TEST(name) static void test_##name(void)
#define RUN_TEST(name) do { \
    printf("  Running %s... ", #name); \
    tests_run++; \
    test_##name(); \
    tests_passed++; \
    printf("PASSED\n"); \
} while(0)

#define ASSERT_EQ(expected, actual) do { \
    if ((expected) != (actual)) { \
        printf("FAILED at line %d: expected %d, got %d\n", __LINE__, (int)(expected), (int)(actual)); \
        return; \
    } \
} while(0)

#define ASSERT_TRUE(cond) do { \
    if (!(cond)) { \
        printf("FAILED at line %d: condition not true\n", __LINE__); \
        return; \
    } \
} while(0)

#define ASSERT_FALSE(cond) do { \
    if (cond) { \
        printf("FAILED at line %d: condition not false\n", __LINE__); \
        return; \
    } \
} while(0)

/* ============== Quality Code Tests ============== */

TEST(quality_codes_correct_values)
{
    /* Verify quality codes match OPC UA specification */
    ASSERT_EQ(0x00, QUALITY_GOOD);
    ASSERT_EQ(0x40, QUALITY_UNCERTAIN);
    ASSERT_EQ(0x80, QUALITY_BAD);
    ASSERT_EQ(0xC0, QUALITY_NOT_CONNECTED);
}

TEST(quality_is_usable_for_control)
{
    /* GOOD and UNCERTAIN quality can be used for control */
    ASSERT_TRUE(QUALITY_GOOD == 0x00 || QUALITY_GOOD == 0x40);
    ASSERT_TRUE(QUALITY_UNCERTAIN == 0x40);

    /* BAD and NOT_CONNECTED should not be used */
    ASSERT_TRUE(QUALITY_BAD >= 0x80);
    ASSERT_TRUE(QUALITY_NOT_CONNECTED >= 0x80);
}

/* ============== Sensor Data Quality Tests ============== */

TEST(sensor_data_includes_quality)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_GOOD;
    sensor.timestamp_ms = 1000;
    sensor.stale = false;

    /* Verify structure contains all quality fields */
    ASSERT_EQ(IOPS_GOOD, sensor.status);
    ASSERT_EQ(QUALITY_GOOD, sensor.quality);
    ASSERT_FALSE(sensor.stale);
}

TEST(sensor_stale_detection)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_GOOD;
    sensor.timestamp_ms = 1000;
    sensor.stale = false;

    /* Simulate staleness check (would be done by registry) */
    uint64_t now = 10000;  /* 9 seconds later */
    uint64_t age = now - sensor.timestamp_ms;

    if (age > 5000) {
        sensor.stale = true;
        if (sensor.quality == QUALITY_GOOD) {
            sensor.quality = QUALITY_UNCERTAIN;
        }
    }

    ASSERT_TRUE(sensor.stale);
    ASSERT_EQ(QUALITY_UNCERTAIN, sensor.quality);
}

TEST(sensor_extended_stale_becomes_bad)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_GOOD;
    sensor.timestamp_ms = 1000;

    /* Simulate 90 seconds stale (>60s threshold) */
    uint64_t now = 91000;
    uint64_t age = now - sensor.timestamp_ms;

    if (age > 60000) {
        sensor.stale = true;
        sensor.quality = QUALITY_BAD;
    }

    ASSERT_TRUE(sensor.stale);
    ASSERT_EQ(QUALITY_BAD, sensor.quality);
}

TEST(sensor_very_stale_becomes_not_connected)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_GOOD;
    sensor.timestamp_ms = 1000;

    /* Simulate 6 minutes stale (>5min threshold) */
    uint64_t now = 361000;
    uint64_t age = now - sensor.timestamp_ms;

    if (age > 300000) {
        sensor.stale = true;
        sensor.quality = QUALITY_NOT_CONNECTED;
    }

    ASSERT_TRUE(sensor.stale);
    ASSERT_EQ(QUALITY_NOT_CONNECTED, sensor.quality);
}

/* ============== Control Decision Tests ============== */

TEST(pid_should_reject_bad_quality)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_BAD;

    /* Simulate PID quality check */
    bool quality_ok = (sensor.status == IOPS_GOOD &&
                       (sensor.quality == QUALITY_GOOD ||
                        sensor.quality == QUALITY_UNCERTAIN));

    ASSERT_FALSE(quality_ok);
}

TEST(pid_should_accept_good_quality)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_GOOD;

    /* Simulate PID quality check */
    bool quality_ok = (sensor.status == IOPS_GOOD &&
                       (sensor.quality == QUALITY_GOOD ||
                        sensor.quality == QUALITY_UNCERTAIN));

    ASSERT_TRUE(quality_ok);
}

TEST(pid_should_accept_uncertain_with_warning)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_UNCERTAIN;

    /* Simulate PID quality check - UNCERTAIN is acceptable but logged */
    bool quality_ok = (sensor.status == IOPS_GOOD &&
                       (sensor.quality == QUALITY_GOOD ||
                        sensor.quality == QUALITY_UNCERTAIN));

    bool should_warn = (sensor.quality == QUALITY_UNCERTAIN);

    ASSERT_TRUE(quality_ok);
    ASSERT_TRUE(should_warn);
}

/* ============== Interlock Fail-Safe Tests ============== */

TEST(interlock_trips_on_bad_quality)
{
    sensor_data_t sensor = {0};
    sensor.value = 50.0f;  /* Normal value, not exceeding threshold */
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_BAD;

    interlock_t interlock = {0};
    interlock.enabled = true;
    interlock.condition = INTERLOCK_CONDITION_ABOVE;
    interlock.threshold = 100.0f;

    /* Simulate interlock quality check */
    bool input_valid = (sensor.status == IOPS_GOOD &&
                        sensor.quality == QUALITY_GOOD);

    bool condition_met = false;
    if (input_valid) {
        condition_met = (sensor.value > interlock.threshold);
    } else {
        /* Fail-safe: treat bad input as trip */
        condition_met = true;
    }

    ASSERT_FALSE(input_valid);
    ASSERT_TRUE(condition_met);  /* Should trip despite normal value */
}

TEST(interlock_trips_on_not_connected)
{
    sensor_data_t sensor = {0};
    sensor.value = 50.0f;
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_NOT_CONNECTED;

    interlock_t interlock = {0};
    interlock.enabled = true;
    interlock.condition = INTERLOCK_CONDITION_ABOVE;
    interlock.threshold = 100.0f;

    bool input_valid = (sensor.status == IOPS_GOOD &&
                        sensor.quality == QUALITY_GOOD);

    bool condition_met = !input_valid;  /* Fail-safe */

    ASSERT_FALSE(input_valid);
    ASSERT_TRUE(condition_met);
}

TEST(interlock_normal_on_good_quality)
{
    sensor_data_t sensor = {0};
    sensor.value = 50.0f;  /* Below threshold */
    sensor.status = IOPS_GOOD;
    sensor.quality = QUALITY_GOOD;

    interlock_t interlock = {0};
    interlock.enabled = true;
    interlock.condition = INTERLOCK_CONDITION_ABOVE;
    interlock.threshold = 100.0f;

    bool input_valid = (sensor.status == IOPS_GOOD &&
                        sensor.quality == QUALITY_GOOD);

    bool condition_met = false;
    if (input_valid) {
        condition_met = (sensor.value > interlock.threshold);
    }

    ASSERT_TRUE(input_valid);
    ASSERT_FALSE(condition_met);  /* Should NOT trip */
}

/* ============== Historian Quality Tests ============== */

TEST(historian_sample_stores_quality)
{
    historian_sample_t sample = {0};
    sample.timestamp_ms = 1000;
    sample.tag_id = 1;
    sample.value = 7.5f;
    sample.quality = QUALITY_GOOD;

    ASSERT_EQ(QUALITY_GOOD, sample.quality);

    /* Change quality and verify */
    sample.quality = QUALITY_BAD;
    ASSERT_EQ(QUALITY_BAD, sample.quality);
}

TEST(quality_should_propagate_from_sensor)
{
    /* Simulate sensor -> historian flow */
    sensor_data_t sensor = {0};
    sensor.value = 7.5f;
    sensor.quality = QUALITY_UNCERTAIN;

    historian_sample_t sample = {0};
    sample.value = sensor.value;
    sample.quality = (uint8_t)sensor.quality;  /* Direct propagation */

    ASSERT_EQ(QUALITY_UNCERTAIN, sample.quality);
}

/* ============== IOPS vs Quality Tests ============== */

TEST(iops_bad_means_bad_quality)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_BAD;  /* Protocol level bad */
    sensor.quality = QUALITY_GOOD;  /* Application level looks good */

    /* Quality check should consider BOTH */
    bool usable = (sensor.status == IOPS_GOOD &&
                   (sensor.quality == QUALITY_GOOD ||
                    sensor.quality == QUALITY_UNCERTAIN));

    ASSERT_FALSE(usable);  /* IOPS_BAD makes it unusable */
}

TEST(iops_good_with_bad_quality_unusable)
{
    sensor_data_t sensor = {0};
    sensor.value = 7.0f;
    sensor.status = IOPS_GOOD;  /* Protocol level good */
    sensor.quality = QUALITY_BAD;  /* Application level bad */

    bool usable = (sensor.status == IOPS_GOOD &&
                   (sensor.quality == QUALITY_GOOD ||
                    sensor.quality == QUALITY_UNCERTAIN));

    ASSERT_FALSE(usable);  /* Application quality makes it unusable */
}

/* ============== Test Runner ============== */

void run_quality_tests(void)
{
    printf("\n=== Data Quality Tests ===\n\n");

    printf("Quality Code Tests:\n");
    RUN_TEST(quality_codes_correct_values);
    RUN_TEST(quality_is_usable_for_control);

    printf("\nSensor Data Quality Tests:\n");
    RUN_TEST(sensor_data_includes_quality);
    RUN_TEST(sensor_stale_detection);
    RUN_TEST(sensor_extended_stale_becomes_bad);
    RUN_TEST(sensor_very_stale_becomes_not_connected);

    printf("\nControl Decision Tests:\n");
    RUN_TEST(pid_should_reject_bad_quality);
    RUN_TEST(pid_should_accept_good_quality);
    RUN_TEST(pid_should_accept_uncertain_with_warning);

    printf("\nInterlock Fail-Safe Tests:\n");
    RUN_TEST(interlock_trips_on_bad_quality);
    RUN_TEST(interlock_trips_on_not_connected);
    RUN_TEST(interlock_normal_on_good_quality);

    printf("\nHistorian Quality Tests:\n");
    RUN_TEST(historian_sample_stores_quality);
    RUN_TEST(quality_should_propagate_from_sensor);

    printf("\nIOPS vs Quality Tests:\n");
    RUN_TEST(iops_bad_means_bad_quality);
    RUN_TEST(iops_good_with_bad_quality_unusable);

    printf("\n=== Results: %d/%d tests passed ===\n\n", tests_passed, tests_run);
}

int main(int argc, char *argv[])
{
    (void)argc;
    (void)argv;
    run_quality_tests();
    return (tests_passed == tests_run) ? 0 : 1;
}
